assets/js/googlesitekit/widgets/datastore/widgets.js

```diff
@@ -176,88 +177,64 @@ export const actions = {
 
 export const controls = {};
 
-export const reducer = ( state, { type, payload } ) => {
+export const reducer = createReducer( ( draft, { type, payload } ) => {
 	switch ( type ) {
 		case ASSIGN_WIDGET: {
 			const { slug, areaSlugs } = payload;
 
-			const { areaAssignments } = state;
 			areaSlugs.forEach( ( areaSlug ) => {
-				if ( areaAssignments[ areaSlug ] === undefined ) {
-					areaAssignments[ areaSlug ] = [];
+				if ( draft.areaAssignments[ areaSlug ] === undefined ) {
+					draft.areaAssignments[ areaSlug ] = [];
 				}
 
-				if ( ! areaAssignments[ areaSlug ].includes( slug ) ) {
-					areaAssignments[ areaSlug ].push( slug );
+				if ( ! draft.areaAssignments[ areaSlug ].includes( slug ) ) {
+					draft.areaAssignments[ areaSlug ].push( slug );
 				}
 			} );
 
-			return {
-				...state,
-				areaAssignments,
-			};
+			return draft;
 		}
 
 		case REGISTER_WIDGET: {
 			const { slug, settings } = payload;
 
-			if ( state.widgets[ slug ] !== undefined ) {
+			if ( draft.widgets[ slug ] !== undefined ) {
 				global.console.warn(
 					`Could not register widget with slug "${ slug }". Widget "${ slug }" is already registered.`
 				);
 
-				return state;
+				return draft;
 			}
 
-			return {
-				...state,
-				widgets: {
-					...state.widgets,
-					[ slug ]: {
-						...settings,
-						slug,
-					},
-				},
-			};
+			draft.widgets[ slug ] = { ...settings, slug };
+			return draft;
 		}
 
 		case SET_WIDGET_STATE: {
 			const { slug, Component, metadata } = payload;
 
-			return {
-				...state,
-				widgetStates: {
-					...state.widgetStates,
-					[ slug ]: {
-						Component,
-						metadata,
-					},
-				},
-			};
+			draft.widgetStates[ slug ] = { Component, metadata };
+			return draft;
 		}
 
 		case UNSET_WIDGET_STATE: {
 			const { slug, Component, metadata } = payload;
 
-			const widgetStates = { ...state.widgetStates };
 			if (
-				widgetStates?.[ slug ]?.Component === Component &&
-				widgetStates?.[ slug ]?.metadata === metadata
+				draft.widgetStates?.[ slug ]?.Component === Component &&
+				draft.widgetStates?.[ slug ]?.metadata === metadata
```

Comment:
This is not obvious at first, I don't blame you for missing it.

Because we are doing a referential comparison of a draft object, we need to use [`original`](https://immerjs.github.io/immer/original/) to extract the original object from the draft in order to compare it.

```suggestion
				original( draft.widgetStates?.[ slug ]?.metadata ) === metadata
```

We _don't_ need to call `original` to extract `Component` because functions are not drafted by default.